### ISSUE TITLE
Saturation of training

### ISSUE THREAD

**Issue by c-j-johnston** [wilsonmr/anvil#18]
I've been trying to imprive acceptance by increasing the number of layers in the `s` and `t` networks (ie change `affine_hidden_shape` from `(32,)` to something like `(32,32)`) but acceptance is actually decreasing. I think this is becuase we are using `nn.ModuleList` when we should be using `nn.Sequential` as per this post: https://discuss.pytorch.org/t/when-should-i-use-nn-modulelist-and-when-should-i-use-nn-sequential/5463/3. I'm going to play about with changing it and let you know if I see improvements.

**Comment by wilsonmr** [comment]
my first response is it won't make a difference, the module list just lets the network know that the modules in the list are layers of the current module, and the gradient should be calculated from the forward pass (which builds a graph of operations) but let me know if I made a mistake

**Comment by c-j-johnston** [comment]
From that post (which I realise is just from a forum and not from the docs), it seems like layers in `ModuleList` aren't connected, while those in `Sequential` are, and page 6 of the paper says: "The neural networks si and ti used in coupling layer gi (see Eq. (9)) were constructed from two to six fully-connected layers". I'll see if it makes a difference to make the change; if not I'll leave the code as is.

**Comment by wilsonmr** [comment]
a fully connected layer is like nn.Linear where every node in one layer is connected to the next as opposed to like a convolution layer or some kind of sparse layer

the modules being in a module list has no relation to their connectedness, the module list just tells the model that model.parameters includes the parameters of a bunch of layers (otherwise I have to declare them one by one)

the modules in module list then become sequential when we do the forward pass because we do like

x -> nn.Linear -> output1 -> nn.Linear -> output 2 -> ... -> output

the above sequence is done using the `__call__` method of the modules in the list which have hooks which include the operations in the DAG (direct acyclic graph) of operations which take input x eventually to loss L. Then pytorch uses this graph to calculate the gradients of all the parameters of the model (which it knows because of the module list)

**Comment by wilsonmr** [comment]
Now having said that nn.Sequential means that in the forward pass we don't have to loop over layers, instead the sequential appears to have a forward pass -> this would definitely look nicer but doesn't affect gradient calcs

**Comment by c-j-johnston** [comment]
I see what you mean, the connectedness comes from the `_s_forward` and `_t_forward` methods further down the script -- I missed that detail -- so we can use `ModuleList`. In that case I'm not sure why my acceptance is so much lower when I add layers. I'm on ~0.5 for (32,), ~0.42 for (32,32) and ~0.21 for (32,32,32,32).

**Comment by wilsonmr** [comment]
well it's not always as simple as adding layers with these things. since we have 6x6=36 layers in the input (and output) I wonder if before making the network deeper you tried making it wider - like single layer (64,) or (128,) or something? could be that the minimiser algorithm isn't the best one - I picked it pretty much at random, you could try changing it to Adam or whatever they use in the paper.

Another thing is the batch size could be we want smaller or larger batches, in fact there are a bunch of things, but maybe start with the bigger single layer suggestion

**Comment by c-j-johnston** [comment]
Ok, I'll give those a go. More layers seemes intuituve but maybe it's not the best parameter to change. Thanks

**Comment by wilsonmr** [comment]
Renamed to heart of the issue - we can't seem to train past a certain point and don't know what to do about it... Yet.

**Comment by jmarshrossney** [comment]
We've figured out why we training plateaued much earlier than expected, the culprits being
- the ReLU on the final layer of the neural networks (specifically, the constant shift **t** networks)
- The need to tune the learning rate and scheduler more carefully

There are still improvements to be made but I think this issue can be closed.
